File: app/src/main/java/com/sysu/edu/widget/WidgetUpdateWorker.java
package com.sysu.edu.widget;

import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;

import androidx.annotation.NonNull;
import androidx.datastore.preferences.core.MutablePreferences;
import androidx.datastore.preferences.core.Preferences;
import androidx.datastore.rxjava3.RxDataStore;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import com.alibaba.fastjson2.JSONArray;
import com.alibaba.fastjson2.JSONObject;
import com.sysu.edu.api.CommonUtil;
import com.sysu.edu.model.JwxtModel;

import java.util.Objects;

import io.reactivex.rxjava3.core.Single;
import okhttp3.Request;

public class WidgetUpdateWorker extends Worker {
    
    final JwxtWidgetModel model = new JwxtWidgetModel(getApplicationContext());
    
    public WidgetUpdateWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }
    
    @NonNull
    @Override
    public Result doWork() {
        try {
            JSONArray networkData = getData();
            RxDataStore<Preferences> dataStore = DataStoreManager.getInstance(getApplicationContext());
            dataStore.updateDataAsync(prefsIn -> {
                MutablePreferences mutablePreferences = prefsIn.toMutablePreferences();
                if (networkData != null)
                    mutablePreferences.set(DataStoreManager.TODAY_CLASS, networkData.toJSONString());
                return Single.just(mutablePreferences);
            });
            String widgetName = getInputData().getString("component");
            String[] widgetNames = getInputData().getStringArray("components");
            if (widgetName != null)
                updateWidget(widgetName);
            else if (widgetNames != null)
                for (String name : widgetNames)
                    updateWidget(name);
            return Result.success();
        } catch (Exception e) {
            return Result.failure();
        }
    }
    
    private void updateWidget(String name) throws ClassNotFoundException {
        updateWidget(Class.forName(getApplicationContext().getPackageName() + ".widget." + name));
    }
    
    private void updateWidget(Class<?> widgetClass) {
        getApplicationContext().sendBroadcast(new Intent(getApplicationContext(), widgetClass)
                .setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE)
                .putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, AppWidgetManager.getInstance(getApplicationContext())
                        .getAppWidgetIds(new ComponentName(getApplicationContext(), widgetClass))));
    }
    
    private JSONArray getData() {
        getTerm();
        CommonUtil.Tuple2<Integer, JSONObject> r1 = model.execute(Objects.requireNonNull(model.getNextRequest()));
        String term = null;
        if (r1 != null) {
            term = r1.second.getJSONObject("data").getString("acadYearSemester");
        }
        getWeek(term);
        CommonUtil.Tuple2<Integer, JSONObject> r2 = model.execute(model.getNextRequest());
        getTodayCourses(term);
        CommonUtil.Tuple2<Integer, JSONObject> r3 = model.execute(model.getNextRequest());
        return (r1 != null && r2 != null && r3 != null)
                ? JSONArray.of(r1.second, r2.second, r3.second) : null;
    }
    
    void getTerm() {
        model.add("jwxt/base-info/acadyearterm/showNewAcadlist", 0);
    }
    
    void getWeek(String term) {
        model.add("jwxt/timetable-search/classTableInfo/getDateWeekly?academicYear=" + term, 1);
    }
    
    void getTodayCourses(String term) {
        model.add("jwxt/timetable-search/classTableInfo/queryTodayStudentClassTable?academicYear=" + term, 2);
    }
    
    static class JwxtWidgetModel extends JwxtModel {
        public JwxtWidgetModel(Context context) {
            super(context);
        }
        
        @Override
        protected void retry(@NonNull CommonUtil.Tuple2<Request, Integer> request) {
            execute(request);
        }
    }
}

File: app/src/main/java/com/sysu/edu/widget/WidgetUpdateWorker.kt
package com.sysu.edu.widget

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import androidx.datastore.preferences.core.Preferences
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.alibaba.fastjson2.JSONArray
import com.alibaba.fastjson2.JSONObject
import com.sysu.edu.api.CommonUtil
import com.sysu.edu.api.DataStoreManager
import com.sysu.edu.model.JwxtModel
import io.reactivex.rxjava3.core.Single
import kotlinx.coroutines.ExperimentalCoroutinesApi
import okhttp3.Request

class WidgetUpdateWorker(context: Context, workerParams: WorkerParameters) :
	Worker(context, workerParams) {
	val model: JwxtWidgetModel = JwxtWidgetModel(applicationContext)
	@OptIn(ExperimentalCoroutinesApi::class) override fun doWork(): Result {
		try {
			val networkData = data
			val dataStore = DataStoreManager.getInstance(applicationContext)
			dataStore.updateDataAsync { prefsIn: Preferences? ->
				val mutablePreferences = prefsIn!!.toMutablePreferences()
				if (networkData != null) mutablePreferences[DataStoreManager.TODAY_CLASS] = networkData.toJSONString()
				Single.just(mutablePreferences)
			}
			val widgetName = inputData.getString("component")
			val widgetNames: Array<String>? = inputData.getStringArray("components")
			if (widgetName != null) updateWidget(widgetName)
			else widgetNames?.forEach {
				updateWidget(it)
			}
			return Result.success()
		} catch (_: Exception) {
			return Result.failure()
		}
	}
	
	@Throws(ClassNotFoundException::class) private fun updateWidget(name: String?) {
		updateWidget(Class.forName(applicationContext.packageName + ".widget." + name))
	}
	
	private fun updateWidget(widgetClass: Class<*>) {
		applicationContext.sendBroadcast(Intent(applicationContext, widgetClass).setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE)
											 .putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, AppWidgetManager.getInstance(applicationContext)
												 .getAppWidgetIds(ComponentName(applicationContext, widgetClass))))
	}
	
	private val data: JSONArray?
		get() {
			term
			val r1: CommonUtil.Tuple2<Int, JSONObject>? = model.execute(model.nextRequest!!)
			val term: String? = r1?.second?.getJSONObject("data")?.getString("acadYearSemester")
			getWeek(term)
			val r2: CommonUtil.Tuple2<Int, JSONObject>? = model.execute(model.nextRequest!!)
			getTodayCourses(term)
			val r3: CommonUtil.Tuple2<Int, JSONObject>? = model.execute(model.nextRequest!!)
			return if (r2 != null && r3 != null) JSONArray.of(r1?.second, r2.second, r3.second) else null
		}
	val term: Unit
		get() {
			model.add("jwxt/base-info/acadyearterm/showNewAcadlist", 0)
		}
	
	fun getWeek(term: String?) {
		model.add("jwxt/timetable-search/classTableInfo/getDateWeekly?academicYear=$term", 1)
	}
	
	fun getTodayCourses(term: String?) {
		model.add("jwxt/timetable-search/classTableInfo/queryTodayStudentClassTable?academicYear=$term", 2)
	}
	
	class JwxtWidgetModel(context: Context) : JwxtModel(context) {
		override fun retry(request: CommonUtil.Tuple2<Request, Int>) {
			execute(request)
		}
	}
}

